swarm/src/connection/handler_wrapper.rs

```diff
@@ -440,3 +440,84 @@ pub enum Event<TOutboundOpenInfo, TCustom> {
     /// Other event.
     Custom(TCustom),
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::handler::PendingConnectionHandler;
+    use quickcheck::*;
+    use std::num::NonZeroU8;
+    use std::sync::Arc;
+
+    struct DummySubstream(Arc<()>);
+
+    impl AsyncRead for DummySubstream {
+        fn poll_read(
+            self: Pin<&mut Self>,
+            _cx: &mut Context<'_>,
+            _buf: &mut [u8],
+        ) -> Poll<std::io::Result<usize>> {
+            Poll::Pending
+        }
+    }
+
+    impl AsyncWrite for DummySubstream {
+        fn poll_write(
+            self: Pin<&mut Self>,
+            _cx: &mut Context<'_>,
+            _buf: &[u8],
+        ) -> Poll<std::io::Result<usize>> {
+            Poll::Pending
+        }
+
+        fn poll_flush(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<std::io::Result<()>> {
+            Poll::Pending
+        }
+
+        fn poll_close(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<std::io::Result<()>> {
+            Poll::Pending
+        }
+    }
+
+    #[test]
+    fn max_negotiating_inbound_streams() {
+        fn prop(max_negotiating_inbound_streams: NonZeroU8) {
```

Comment:
This is only a `NonZeroU8` so that quickcheck does not generate `0`?

Reply:
If the zero case is not interesting, shouldn't we also use a `NonZero` type in `HandlerWrapper`?

Reply:
Good point. I think we should support the use-case of `0`. That is, I think we should support the use-case of blocking all inbound streams.

With that in mind, the test should cover it. https://github.com/libp2p/rust-libp2p/pull/2785/commits/87b9096bf9d987cb366909b93469cb55144579a8 does the adjustments.

Thanks @thomaseizinger.